Add ability to filter listings based on jobs

calvinjhchan opened the pull request:
![image](https://github.com/user-attachments/assets/b8ac947b-26f8-4153-9ec4-e8d1c9cd9457)

New to all this, especially Rust, so open to any suggestions. Tried my best to match existing code/style.

Implementation:
- For each listing, store all joinable roles/jobs using bitflags
- In the filter, each job has a checkbox that toggles their bitflag
- Then it is compared to the joinable roles to determine if the listing should be shown. Note that I had to use BigInt since javascript uses 32-bit signed integers for bitwise operations, which was resulting in negatives.

Missing:
- I only rendered the jobs to avoid clutter, but it shouldn't be too hard to add classes if we really want
- ~~Doesn't really work for alliance raids or pvp since we don't have accurate data for jobs present. Maybe I can have the filter ignore those.~~
  - ^ This is done. The filter only acts on listings with 1 party
- ~~Do we need translations?~~
- ~~Retain state between refreshes~~
- Tests

zeroeightysix commented:
Hi, thanks for working on this! I'll review this in detail soon, but looks quite good already from a (very quick) scan of the changes :)

zeroeightysix commented on server/src/listing.rs:
```diff
@@ -72,6 +73,34 @@ impl PartyFinderListing {
+            .contains(SearchAreaFlags::ONE_PLAYER_PER_JOB);
+        let mut jobs = JobFlags::empty();
+        let mut jobs_taken = JobFlags::empty();
+        for i in 0..self.slots_available as usize {
+            if i >= self.jobs_present.len() {
+                break;
+            }
+
+            match JOBS.get(&u32::from(self.jobs_present[i])) {
```
Realistically, `jobs_present` should always be smaller or equal in size as `slots_available`: so why not iterate over its indices instead?

```suggestion
    for (i, present_job) in self.jobs_present.iter().copied().enumerate() {
            match JOBS.get(&(present_job as u32)) {
```

Because you're also using `i` to index into `slots`, a bounds check on that would be appropriate though (although that edge case would rarely hit, I'd much rather the filter misbehave than the server panic :) )

calvinjhchan replied:
`slots_available` can be smaller than `jobs_present` length (which is always 8 from what I see) if slots are omitted, but yeah, iterating over `jobs_present` should be cleaner

zeroeightysix commented:
Also, now that the job filters take up more vertical space than the category filter, would you be up for it to make the category filter grow to the bottom of the available size?

![image](https://github.com/user-attachments/assets/8b719737-4854-49bb-bbef-c455f3ee1560)

calvinjhchan commented:
![image](https://github.com/user-attachments/assets/3d50eda3-c9a2-4a56-b879-4c886c465b2d)

Addressed comments and fixed some formatting. Any thoughts for the items under "missing"?

zeroeightysix commented:
> Addressed comments and fixed some formatting. Any thoughts for the items under "missing"?

Thanks!

* People are only filtering for jobs 99% of the time, and I do think adding classes would be a lot of visual clutter. I'm fine leaving it out.
* We don't? Ignoring is fine, though.
* Yes, if you can add translations, that would be appreciated.
* The current existing tests don't even compile, so honestly.. don't bother for the time being :-)

calvinjhchan commented:
> We don't? Ignoring is fine, though.

Yeah, I noticed that the slots filled can show 0 and accepting any role, even if they are restricted:
![image](https://github.com/user-attachments/assets/123eb518-5011-4bcf-91b2-f86607f751d7)
![image](https://github.com/user-attachments/assets/e58f5f89-8fc0-421d-b118-2e793c3e89a5)

calvinjhchan commented:
![image](https://github.com/user-attachments/assets/4a6a5d86-a71e-49b2-be86-61cf8a90a674)
![image](https://github.com/user-attachments/assets/9e8be005-5e32-4666-8809-83146d5b2801)

The auto translate entries didn't have the exact translations we needed, so I pulled them from the job guide instead (https://jp.finalfantasyxiv.com/jobguide/battle/)

zeroeightysix commented:
Looks great. Thanks again!

zeroeightysix commented:
Deploying this revealed a bit of an issue I'll have to debug later. `<input>`s in filters aren't rendering their icon, which doesn't happen when I run it locally. Quite weird, but I'll see if I can figure it out tomorrow :)

calvinjhchan commented:
I noticed the js/css assets were missing the latest changes

zeroeightysix commented:
> I noticed the js/css assets were missing the latest changes

Oops. Feel free to commit those to your branch, and I'll merge them as well

calvinjhchan commented:
Sorry, I meant that when the changes were briefly deployed, the website was still sending the old js/css files.

zeroeightysix commented:
> Sorry, I meant that when the changes were briefly deployed, the website was still sending the old js/css files.

Good catch. I've fixed the issue and it's now deployed. Thank you!
